sdk/core/azure-core/inc/http/policy.hpp

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "context.hpp"
+#include "http.hpp"
+
+namespace Azure
+{
+namespace Core
+{
+namespace Http
+{
+
+class HttpPolicy
+{
+public:
+  // If we get a response that goes up the stack
+  // Any errors in the pipeline throws an exception
+  // At the top of the pipeline we might want to turn certain responses into exceptions
+  virtual Response Process(Context& context, Request& request) const = 0;
+  virtual ~HttpPolicy() {}
+};
+
+class HttpTransport : public HttpPolicy
+{
+  Response Process(Context& context, Request& request) const override
+  {
+    context.CancelWhen();
+    request.getHeaders();
+
+    return Response(200, "OK\n");
+  }
+};
+
+struct RetryPolicyOptions
+{
+
+private:
+  int16_t max_retries;
+  int32_t retry_delay_msec;
+
+public:
+  RetryPolicyOptions()
+  {
+    max_retries = 5;
+    retry_delay_msec = 500;
+  };
+};
+
+class RetryPolicy : public HttpPolicy
+{
+private:
+  std::unique_ptr<HttpPolicy> nextPolicy_;
```

Comment:
Naming not fixed.

Reply:
Oh wait, we do NOT want to construct policies passing to them a 'nextPolicy' as this forms a linked list.
Instead, we want to do the same way we did it in C. Each policy object has its options and then the calling code creates the array of policy objects and then calls Processes passing in the array. We call the next policy's Process by walking to the next element in the array.

What I say is more efficient: We can create just 1 retry policy object, configure it and use it from multiple pipelines that have different policies simultaneously.